go/database/mpt/tool/stress.go

```diff
@@ -0,0 +1,348 @@
+	const (
+		MiB       = 1024 * 1024
+		cacheSize = 64 * MiB
+	)
+
+	tmpDir := context.String(tmpDirFlag.Name)
+	if len(tmpDir) == 0 {
+		tmpDir = os.TempDir()
+	}
+
+	dir := filepath.Join(tmpDir, fmt.Sprintf("carmen-stress-%d", time.Now().UnixNano()))
+	fmt.Printf("Using temporary directory: %s\n", dir)
+
+	flushPeriod := context.Duration(flushPeriodFlag.Name)
+	fmt.Printf("Using background flush period: %s\n", flushPeriod)
+
+	reportPeriod := context.Duration(reportPeriodFlag.Name)
+	fmt.Printf("Using report period: %s\n", reportPeriod)
+
+	cacheConfig := mpt.NodeCacheConfig{
+		Capacity:              cacheSize / mpt.EstimatePerNodeMemoryUsage(),
+		BackgroundFlushPeriod: flushPeriod,
+	}
+
+	db, err := mpt.OpenGoFileState(dir, mpt.S5LiveConfig, cacheConfig)
+	if err != nil {
+		return fmt.Errorf("failed to open database: %w", err)
+	}
+
+	numBlocks := context.Int(numBlocksFlag.Name)
+	if numBlocks <= 0 {
+		numBlocks = 1000
+	}
+	fmt.Printf("Inserting %d blocks ...\n", numBlocks)
+
+	state := createTestState(db, dir)
+
+	var reportWg sync.WaitGroup
+	reportWg.Add(1)
+	stopReport := make(chan struct{})
+	defer func() {
+		close(stopReport)
+		reportWg.Wait()
+	}()
+	go func() {
+		defer reportWg.Done()
+		ticker := time.NewTicker(reportPeriod)
+		for {
+			select {
+			case <-stopReport:
+				return
+			case <-ticker.C:
+				state.ReportProgress()
+			}
+		}
+	}()
+
+	stopRun := make(chan os.Signal, 1)
+	signal.Notify(stopRun, syscall.SIGINT, syscall.SIGTERM)
+
+	rand := rand.New(rand.NewSource(state.start.UnixNano()))
+loop:
+	for i := 0; i < numBlocks; i++ {
+		select {
+		case <-stopRun:
+			fmt.Printf("Stopped by interrupt signal ...\n")
+			break loop
+		default:
+		}
+		if err := state.AddBlock(rand); err != nil {
+			return fmt.Errorf("failed to add block %d: %w", i, err)
+		}
+	}
+
+	if err := db.Close(); err != nil {
+		return fmt.Errorf("failed to close database: %w", err)
+	}
+
+	if err := os.RemoveAll(dir); err != nil {
+		return fmt.Errorf("failed to remove directory: %w", err)
+	}
+
+	return nil
+}
+
+// --- Stress Test State ---
+type stressTestState struct {
+	start       time.Time
+	directory   string
+	db          *mpt.MptState
+	lock        sync.Mutex
+	state       map[int]map[int]int
+	blockHeight int
+	nextAccount int
+	nextKey     int
+}
+
+func createTestState(db *mpt.MptState, directory string) *stressTestState {
+	return &stressTestState{
+		directory: directory,
+		db:        db,
+		start:     time.Now(),
+		state:     map[int]map[int]int{},
+	}
+}
+
+func (s *stressTestState) ReportProgress() {
+	memUsage := getMemoryUsage()
+	used := getDirectorySize(s.directory)
+	free, err := getFreeSpace(s.directory)
+	if err != nil {
+		fmt.Printf("failed to get free space: %v\n", err)
+		return
+	}
+
+	s.lock.Lock()
+	numAccounts := len(s.state)
+	numSlots := 0
+	for _, storage := range s.state {
+		numSlots += len(storage)
+	}
+	currentBlock := s.blockHeight
+	s.lock.Unlock()
+
+	time := time.Since(s.start)
```

Comment:
`Duration` has methods `Hours`, `Minutes`, `Seconds`

Reply:
Yes, but they are not very useful for formatting the time. (a) they produce fractions and (b) you still need to remove the full hours from the minutes and the full minutes from the seconds to get the values I want to print here.

Reply:
I see, thanks for explanation